File: apps/web/lib/shield-service.ts
import { supabase } from './supabase-client';
import { getTodayLocal, toLocalDayString } from './daily-service';

/**
 * Check if the player's streak is at risk (>20h since last play or missed a day).
 */
export function isStreakAtRisk(lastPlayedAt: string | null): boolean {
  if (!lastPlayedAt) return false;

  const last = new Date(lastPlayedAt);
  const now = new Date();
  const hoursSince = (now.getTime() - last.getTime()) / (1000 * 60 * 60);

  // At risk if >20h since last play AND the calendar day has rolled over
  // in the player's local timezone (matches our daily reset boundary).
  if (hoursSince > 20) {
    return toLocalDayString(last) !== getTodayLocal();
  }

  return false;
}

/**
 * Use a shield to protect the streak. Returns true if successful.
 */
export async function useShield(userId: string): Promise<boolean> {
  const { data: profile } = await (supabase as any)
    .from('profiles')
    .select('streak_shields, daily_login_streak, last_played_at')
    .eq('id', userId)
    .single() as { data: { streak_shields: number; daily_login_streak: number; last_played_at: string | null } | null };

  if (!profile || profile.streak_shields <= 0) return false;

  // Update shield count and preserve streak by updating last_played_at to now
  await (supabase as any)
    .from('profiles')
    .update({
      streak_shields: profile.streak_shields - 1,
      last_played_at: new Date().toISOString(),
    })
    .eq('id', userId);

  return true;
}

/**
 * Grant the 7-day streak-milestone shield.
 *
 * Delegated to /api/shields/grant-milestone. Shields are a paid benefit (+4 on
 * every Pro renewal) and a referral reward, and nothing in the DB stopped a
 * client from simply writing itself more — so the grant can't live here. The
 * server recomputes the streak from daily_results and keeps its own
 * already-paid ledger, making replays worthless.
 */
export async function grantFreeShield(_userId: string): Promise<void> {
  const { data: { session } } = await supabase.auth.getSession();
  if (!session?.access_token) return;
  try {
    await fetch('/api/shields/grant-milestone', {
      method: 'POST',
      headers: { Authorization: `Bearer ${session.access_token}` },
    });
  } catch {
    // Best-effort: a missed milestone shield must never break result recording.
    // The next call re-derives it from the ledger, so nothing is lost.
  }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/ShieldService.kt
package com.wordocious.app.data

import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Streak-shield logic — Android port of web lib/shield-service.ts.
 * A streak is "at risk" when >20h have passed since last play AND the local
 * calendar day rolled over; using a shield decrements streak_shields and
 * refreshes last_played_at so the streak survives.
 */
object ShieldService {
    private val client get() = SupabaseConfig.client

    /**
     * Claim the 7-day-streak milestone shield.
     *
     * Server-owned: streak_shields is a paid benefit (+4 per billing period) and
     * a referral reward, so a client that can raise it can mint one for free.
     * The endpoint recomputes the streak from daily_results and keeps its own
     * already-paid ledger, so calling it twice grants nothing.
     */
    suspend fun grantMilestoneShield() {
        runCatching {
            val token = client.auth.currentSessionOrNull()?.accessToken ?: return
            withContext(Dispatchers.IO) {
                val conn = (java.net.URL("https://wordocious.com/api/shields/grant-milestone")
                    .openConnection() as java.net.HttpURLConnection).apply {
                    requestMethod = "POST"
                    setRequestProperty("Authorization", "Bearer $token")
                    connectTimeout = 10_000
                    readTimeout = 10_000
                }
                conn.responseCode
                conn.disconnect()
            }
        }
    }

    fun isStreakAtRisk(lastPlayedAt: String?): Boolean {
        if (lastPlayedAt == null) return false
        return runCatching {
            // OffsetDateTime: PostgREST serializes timestamptz as "+00:00",
            // which Instant.parse rejects on API 26-29 (silent streak resets).
            val last = java.time.OffsetDateTime.parse(
                if (lastPlayedAt.endsWith("Z") || lastPlayedAt.contains('+')) lastPlayedAt else lastPlayedAt + "Z",
            ).toInstant()
            val hoursSince = java.time.Duration.between(last, java.time.Instant.now()).toMinutes() / 60.0
            if (hoursSince > 20) {
                val lastLocalDay = last.atZone(java.time.ZoneId.systemDefault()).toLocalDate().toString()
                lastLocalDay != com.wordocious.app.todayLocalDate()
            } else {
                false
            }
        }.getOrDefault(false)
    }

    @Serializable
    private data class ShieldRow(
        @SerialName("streak_shields") val shields: Int = 0,
    )

    /** Spend one shield to preserve the streak. Returns true on success. */
    suspend fun useShield(userId: String): Boolean = runCatching {
        val p = client.postgrest["profiles"]
            .select(Columns.raw("streak_shields")) { filter { eq("id", userId) }; limit(1) }
            .decodeSingleOrNull<ShieldRow>() ?: return false
        if (p.shields <= 0) return false
        client.postgrest["profiles"].update({
            set("streak_shields", p.shields - 1)
            set("last_played_at", java.time.Instant.now().toString())
        }) { filter { eq("id", userId) } }
        true
    }.getOrDefault(false)

    /** Decline: let the streak reset (web sets daily_login_streak = 0). */
    suspend fun declineStreak(userId: String) {
        runCatching {
            client.postgrest["profiles"].update({ set("daily_login_streak", 0) }) {
                filter { eq("id", userId) }
            }
        }
    }
}
